File: cpp/src/core/state/app_state.cpp
#include "pipela/core/state/app_state.hpp"

#include <unordered_map>

namespace pipela::core::state {

namespace {

using GetterFn = StateValue (*)(const AppState&);
using SetterFn = bool (*)(AppState&, const StateValue&);

template <typename Group, typename Field>
GetterFn makeGetter(Group AppState::* group, Field Group::* field) {
    return [group, field](const AppState& s) -> StateValue { return StateValue{(s.*group).*field}; };
}

template <typename Group, typename Field>
SetterFn makeSetter(Group AppState::* group, Field Group::* field) {
    return [group, field](AppState& s, const StateValue& v) -> bool {
        if (const auto* p = std::get_if<Field>(&v)) {
            (s.*group).*field = *p;
            return true;
        }
        return false;
    };
}

struct KeyOps {
    GetterFn get;
    SetterFn set;
};

const std::unordered_map<std::string, KeyOps>& keyOps() {
    static const std::unordered_map<std::string, KeyOps> table = {
        {"left_click_feature_enabled",
         {makeGetter(&AppState::input, &InputState::left_click_feature_enabled),
          makeSetter(&AppState::input, &InputState::left_click_feature_enabled)}},
        {"left_click_active",
         {makeGetter(&AppState::input, &InputState::left_click_active),
          makeSetter(&AppState::input, &InputState::left_click_active)}},
        {"left_pressed",
         {makeGetter(&AppState::input, &InputState::left_pressed),
          makeSetter(&AppState::input, &InputState::left_pressed)}},
        {"left_click_id",
         {makeGetter(&AppState::input, &InputState::left_click_id),
          makeSetter(&AppState::input, &InputState::left_click_id)}},
        {"flame_trigger_active",
         {makeGetter(&AppState::input, &InputState::flame_trigger_active),
          makeSetter(&AppState::input, &InputState::flame_trigger_active)}},
        {"reload_active",
         {makeGetter(&AppState::input, &InputState::reload_active),
          makeSetter(&AppState::input, &InputState::reload_active)}},
        {"ammo_restock_active",
         {makeGetter(&AppState::input, &InputState::ammo_restock_active),
          makeSetter(&AppState::input, &InputState::ammo_restock_active)}},
        {"running",
         {makeGetter(&AppState::worker, &WorkerRuntimeState::running),
          makeSetter(&AppState::worker, &WorkerRuntimeState::running)}},
        {"target_hwnd",
         {[](const AppState& s) -> StateValue {
              return StateValue{static_cast<std::int64_t>(s.worker.target_hwnd)};
          },
          [](AppState& s, const StateValue& v) -> bool {
              if (const auto* i = std::get_if<int>(&v)) {
                  s.worker.target_hwnd = *i;
                  return true;
              }
              if (const auto* l = std::get_if<std::int64_t>(&v)) {
                  s.worker.target_hwnd = static_cast<std::intptr_t>(*l);
                  return true;
              }
              return false;
          }}},
        {"select_mode",
         {makeGetter(&AppState::worker, &WorkerRuntimeState::select_mode),
          makeSetter(&AppState::worker, &WorkerRuntimeState::select_mode)}},
        {"reload_success_count",
         {makeGetter(&AppState::worker, &WorkerRuntimeState::reload_success_count),
          makeSetter(&AppState::worker, &WorkerRuntimeState::reload_success_count)}},
        {"ammo_restock_loop_count",
         {makeGetter(&AppState::worker, &WorkerRuntimeState::ammo_restock_loop_count),
          makeSetter(&AppState::worker, &WorkerRuntimeState::ammo_restock_loop_count)}},
        {"kill_counter_enabled",
         {makeGetter(&AppState::kill_counter, &KillCounterState::kill_counter_enabled),
          makeSetter(&AppState::kill_counter, &KillCounterState::kill_counter_enabled)}},
        {"kill_counter_session_carried_kills",
         {makeGetter(&AppState::kill_counter, &KillCounterState::kill_counter_session_carried_kills),
          makeSetter(&AppState::kill_counter, &KillCounterState::kill_counter_session_carried_kills)}},
        {"flame_trigger_session_reload_count",
         {makeGetter(&AppState::input, &InputState::flame_trigger_session_reload_count),
          makeSetter(&AppState::input, &InputState::flame_trigger_session_reload_count)}},
    };
    return table;
}

}  // namespace

bool AppState::has(const std::string& key) const { return keyOps().count(key) > 0; }

std::optional<StateValue> AppState::get(const std::string& key) const {
    const auto it = keyOps().find(key);
    if (it == keyOps().end()) {
        return std::nullopt;
    }
    return it->second.get(*this);
}

bool AppState::set(const std::string& key, const StateValue& value) {
    const auto it = keyOps().find(key);
    if (it == keyOps().end()) {
        return false;
    }
    return it->second.set(*this, value);
}

int AppState::incrementInt(const std::string& key, int delta) {
    auto cur = get(key);
    int base = 0;
    if (cur) {
        if (const auto* i = std::get_if<int>(&*cur)) {
            base = *i;
        } else if (const auto* l = std::get_if<std::int64_t>(&*cur)) {
            base = static_cast<int>(*l);
        }
    }
    const int next = base + delta;
    set(key, StateValue{next});
    return next;
}

void AppState::seedFromDefaults() {
    input = InputState{};
    worker = WorkerRuntimeState{};
    kill_counter = KillCounterState{};
}

}  // namespace pipela::core::state

File: pipela_core/app_state.py
"""AGENT: lightweight app state container for phased `main.py` migration."""

from __future__ import annotations

from dataclasses import dataclass, fields
from threading import RLock
from typing import Any


@dataclass
class InputState:
    left_click_feature_enabled: bool
    left_click_active: bool
    left_pressed: bool
    right_hold_active: bool
    left_click_id: int
    flame_trigger_active: bool
    reload_active: bool
    ammo_restock_active: bool
    reload_nobullet_arm_until_mono: float
    ammo_restock_toggle_key_code: int
    flame_trigger_start_time: float
    flame_trigger_press_text_until: float
    flame_trigger_press_key_name: str
    flame_trigger_press_count: int
    flame_trigger_last_press_interval_sec: float
    flame_trigger_prev_press_timestamp: Any
    flame_trigger_hud_session_start_time: float
    flame_trigger_session_reload_count: int
    flame_trigger_last_reload_complete_time: float
    flame_trigger_last_reload_trigger_time: float
    flame_trigger_reload_teardown_preserve_hud: bool
    hp_refill_detection_score: float
    hp_refill_trigger_total: int


@dataclass
class WorkerRuntimeState:
    running: bool
    target_hwnd: Any
    select_mode: bool
    nobullet_detected: bool
    last_nobullet_time: float
    nobullet_detection_score: float
    bullet_detection_score: float
    vault_detection_score: float
    reload_success_count: int
    reload_ammo_count: int
    ammo_restock_loop_count: int
    ammo_restock_buybutton_score: float
    ammo_restock_inven_score: float
    ammo_restock_bank_score: float
    ammo_restock_sequence_busy: bool
    call_merc_sequence_busy: bool
    call_merc_1_score: float
    call_merc_2_score: float
    call_merc_3_score: float
    call_merc_4_score: float
    call_merc_loop_count: int
    start_game_launcher_score: float
    start_game_intro_skip_score: float
    start_game_accept_score: float
    start_game_launcher_loop_count: int


@dataclass
class KillCounterState:
    kill_counter_enabled: bool
    kill_counter_detect_region: Any
    kill_counter_last_progress: str
    kill_counter_last_poll_ts: float
    kill_counter_last_poll_phase: Any
    kill_counter_last_poll_detail: Any
    kill_counter_session_baseline_n1: Any
    kill_counter_session_last_n1: Any
    kill_counter_session_carried_kills: int
    kill_counter_loop_count: int
    _kill_counter_last_change_probe_bgr: Any


@dataclass
class AppState:
    input: InputState
    worker: WorkerRuntimeState
    kill_counter: KillCounterState

    def __post_init__(self) -> None:
        self._lock = RLock()
        self._index: dict[str, tuple[object, str]] = {}
        for grp in (self.input, self.worker, self.kill_counter):
            for f in fields(grp):
                self._index[f.name] = (grp, f.name)

    def has(self, key: str) -> bool:
        return key in self._index

    def get(self, key: str, default: Any = None) -> Any:
        with self._lock:
            if key not in self._index:
                return default
            obj, attr = self._index[key]
            return getattr(obj, attr)

    def set(self, key: str, value: Any) -> None:
        with self._lock:
            if key not in self._index:
                return
            obj, attr = self._index[key]
            setattr(obj, attr, value)

